drivers/usb/udc/udc_dwc2.c

```diff
@@ -2571,7 +2572,11 @@ static void dwc2_handle_incompisoin(const struct device *dev)
 
 				buf = udc_buf_get(dev, cfg->addr);
 				if (buf) {
+					/* Data is no longer relevant */
 					udc_submit_ep_event(dev, buf, 0);
+
+					/* Try to queue next packet before SOF */
```

Comment:
I just found out that there is a race condition between the host, DWC2 otg core and SW. After incomplete iso IN is triggered, it is still possible for the IN to be sent and actually handled. The failed race condition here that results in double pulling data from net buf is as follows:
1. Incomplete iso IN is raised
2. `if ((diepctl & mask) == val)` condition above is true, driver takes path to call `udc_dwc2_ep_disable()`
3. Before `udc_dwc2_ep_disable()` finishes, the host sends the IN token, IN transfer complete interrupt is raised
4. Driver executes `udc_dwc2_ep_disable()`
5. Driver submits next transfer here
6. Host sends SOF
7. IN transfer complete from point 3 is handled but it was for the previous transfer! The net buf is pulled the last tx data
9. Host sends IN token, device responds and IN transfer complete interrupt is raised
10. Driver on next isr loop (without reentering ISR, there is a while() in ISR handler until all interrupts are handled) notices IN transfer complete. The net buf is pulled the last tx data (second time, which triggers assertion failure)

Reply:
The workaround is to clear XferCompl interrupt on endpoint disable. This seems to solve not only the IN problem, but also occassional "udc_dwc2: No buffer for ep" for OUT endpoints.